k3s: add airgap images to passthru attributes

rorosen opened the pull request:
## Description of changes

The k3s update script filters the assets of a
corresponding release for airgap images archives
and provides these as passthru attributes of the
k3s derivation. We use zstd archives, as these
offer the best compression ratios and decompression
speed. Furthermore, the `airgapImages` passthru
provides the images archive that matches the host
platform architecture, however, this only works
for aarch64 and x86_64. In addition, a txt file
listing all container images of a release is made
available via a passthru attribute. The airgap
images archives can be combined nicely with the
`services.k3s.images` option, e.g. to pre-provision
k3s nodes for environments without Internet
connectivity.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

CC @NixOS/k3s

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

yajo commented on nixos/modules/services/cluster/k3s/default.nix:
```diff
@@ -351,12 +351,16 @@ in
             finalImageTag = "21.1.2-debian-11-r0";
           })
+
+          config.services.k3s.package.airgapImagesAmd64
```
question: don't you need to specify passthru?

```suggestion
          config.services.k3s.package.passthru.airgapImagesAmd64
```

rorosen replied:
AFAIK this makes no difference. I get the same store path when building `k3s.airgapImagesAmd64` and `k3s.passthru.airgapImagesAmd64`

yajo commented:
Also I think this is a good candidate for a nixos test. Could you add one please?

superherointj commented:
I think, we should not hardcore arch but infer it from the requested system (`stdenv.hostPlatform`).

superherointj commented:
There is also arm (non-64):
![image](https://github.com/user-attachments/assets/1c5a9985-627b-4598-b40c-8a1a9a8d0c1f)

superherointj commented on pkgs/applications/networking/cluster/k3s/update-script.sh:
```diff
@@ -69,6 +70,22 @@ cat > chart-versions.nix.update <<EOF
+IMAGES_URL=https://github.com/k3s-io/k3s/releases/download/v${K3S_VERSION_URLENCODE}
+rm -f images-versions.nix.update
+cat > images-versions.nix.update <<EOF
+{
```
I'd use a `for` here. Expecting list to change.

superherointj commented:
To match upstream support, `arm` is missing. Risc-V may appear soon.

Optionally, we could abstract per arch and not repeat parts. It is up to you.

wrmilling commented:
Push back if this would be unnecessary, but my knee-jerk reaction would be to not have the end users have to select an architecture as part of pulling the airgap images and instead have it package/pull the image based on the hostPlatform/system of the built k3s package. Thoughts?

rorosen commented:
> Push back if this would be unnecessary, but my knee-jerk reaction would be to not have the end users have to select an architecture as part of pulling the airgap images and instead have it package/pull the image based on the hostPlatform/system of the built k3s package. Thoughts?

Yes, I think this would make a lot of sense. I will implement this, use less hard-coded stuff in the builder script and add a test.

rorosen committed:
k3s: add airgap images to passthru attributes

The k3s update script filters the assets of a
corresponding release for airgap images archives
and provides these as passthru attributes of the
k3s derivation. We use zstd archives, as these
offer the best compression ratios and decompression
speed. Furthermore, the `airgapImages` passthru
provides the images archive that matches the host
platform architecture, however, this only works
for aarch64 and x86_64. In addition, a txt file
listing all container images of a release is made
available via a passthru attribute. The airgap
images archives can be combined nicely with the
`services.k3s.images` option, e.g. to pre-provision
k3s nodes for environments without Internet
connectivity.

rorosen committed:
nixos/k3s: add test for airgap images import

Add a nixos test that imports the airgap images
archive and checks whether all expected images are
successfully loaded by k3s.

rorosen commented:
I made a few changes:

- The update script prefetches all zstd airgap images archives that are associated with the release, if more archives are added, they will be added as passthru attributes automatically
- The builder assigns the `airgapImages` passthru based on the hostPlatform/system, though, this only works for `aarch64` and `x86_64`
- There is additionally an `imageList` passthru that lists all images contained in the airgap archives. This could be useful for SBOMs and similar but also helps with testing
-  There is a NixOS test that checks the import of airgap images
- I changed the "versions" file to be a json as this is much easier to create in the update script and spares me from whitespace formatter errors

This started as a simple idea which I thought was easy to implement but escalated quickly. The current implementation assumes very little of how k3s release artifacts are organized and should be able to adapt automatically to changes in the upstream release process. However, it still uses are more or less hard-coded mapping of architecture strings (e.g. `aarch64` <-> `arm64`).

After all I still think this is very useful, but maybe I'm too influenced by my k3s use case...let me know if you have concerns.

superherointj commented on pkgs/applications/networking/cluster/k3s/builder.nix:
```diff
@@ -122,6 +125,39 @@ let
+    else if stdenv.isx86_64 then
+      imagesVersions.${findImagesArchive "amd64"}
+    else
+      abort "k3s: airgap images cannot be found automatially for architecture ${stdenv.hostPlatform.linuxArch}, consider using an image archive with an explicit architecture."
```
```suggestion
      abort "k3s: airgap images cannot be found automatically for architecture ${stdenv.hostPlatform.linuxArch}, consider using an image archive with an explicit architecture."
```
(minor typo)

rorosen replied:
Good spot, thanks

yajo reviewed (approved):
Quite clever, thanks!

Artturin commented on pkgs/applications/networking/cluster/k3s/builder.nix:
```diff
@@ -122,6 +125,39 @@ let
+    else if stdenv.isx86_64 then
+      imagesVersions.${findImagesArchive "amd64"}
+    else
+      abort "k3s: airgap images cannot be found automatically for architecture ${stdenv.hostPlatform.linuxArch}, consider using an image archive with an explicit architecture."
```
FYI don't use abort in nixpkgs unless absolutely needed, use throw instead because that doesn't abort evaluation when evaluating the whole set.

I'll fix it https://www.github.com/NixOS/nixpkgs/pull/341067
